pkg/trie/triedb/nibbles/nibbles.go

```diff
@@ -0,0 +1,365 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package nibbles
+
+import (
+	"slices"
+)
+
+// Single nibble length in bit.
+const BitsPerNibble uint8 = 4
+
+// Number of nibble per byte.
+const NibblesPerByte uint = 2
+
+// Nibble (half a byte).
+const PaddingBitmask uint8 = 0x0F
+
+// Nibble-orientated view onto byte-slice, allowing nibble-precision offsets.
+//
+// This is meant to be an immutable struct. No operations actually change it.
+type Nibbles struct {
+	data   []uint8
+	offset uint
+}
+
+// Construct new [Nibbles] from data and offset
+func NewNibbles(data []byte, offset ...uint) Nibbles {
+	var off uint
+	if len(offset) > 0 {
+		off = offset[0]
+	}
+	return Nibbles{
+		data:   data,
+		offset: off,
+	}
+}
+
+// Construct new [Nibbles] from [NodeKey]
+func NewNibblesFromNodeKey(from NodeKey) Nibbles {
+	return NewNibbles(from.Data, from.Offset)
+}
+
+// Get the nibble at position i.
+func (n Nibbles) At(i uint) uint8 {
+	ix := (n.offset + i) / NibblesPerByte
+	pad := uint8((n.offset + i) % NibblesPerByte) //nolint:gosec
+	return atLeft(pad, n.data[ix])
+}
+
+func atLeft(ix, b uint8) uint8 {
+	if ix == 1 {
+		return b & 0x0F
+	} else {
+		return b >> BitsPerNibble
+	}
+}
+
+// Return object which represents a view on to this slice (further) offset by i nibbles.
+func (n Nibbles) Mid(i uint) Nibbles {
+	return Nibbles{
+		data:   slices.Clone(n.data),
+		offset: n.offset + i,
+	}
+}
+
+// Mask a byte, keeping left nibble.
+func PadLeft(b uint8) uint8 {
+	return b &^ PaddingBitmask
+}
+
+// Mask a byte, keeping right byte.
+func PadRight(b uint8) uint8 {
+	return b & PaddingBitmask
+}
+
+// A trie node prefix, it is the nibble path from the trie root
+// to the trie node.
+// For a node containing no partial key value it is the full key.
+// For a value node or node containing a partial key, it is the full key minus its node partial
+// nibbles (the node key can be split into prefix and node partial).
+// Therefore it is always the leftmost portion of the node key, so its internal representation
+// is a non expanded byte slice followed by a last padded byte representation.
+// The padded byte is an optional padded value.
+type Prefix struct {
+	Key    []byte
+	Padded *byte
+}
+
+func (p Prefix) JoinedBytes() []byte {
+	if p.Padded != nil {
+		return append(p.Key, *p.Padded)
+	}
+	return p.Key
+}
+
+// Return left portion of [Nibbles], if the slice
+// originates from a full key it will be the Prefix of
+// the node.
+func (n Nibbles) Left() Prefix {
+	split := n.offset / NibblesPerByte
+	ix := uint8(n.offset % NibblesPerByte) //nolint:gosec
+	if ix == 0 {
+		return Prefix{Key: n.data[:split]}
+	}
+	padded := PadLeft(n.data[split])
+	return Prefix{Key: slices.Clone(n.data[:split]), Padded: &padded}
+}
+
+func (n Nibbles) Len() uint {
+	return uint(len(n.data))*NibblesPerByte - n.offset
+}
+
+// Advance the view on the slice by i nibbles.
+func (n *Nibbles) Advance(i uint) {
+	if !(n.Len() >= i) {
```

Comment:
```suggestion
	if n.Len() < i {
```
But also, shouldn't this be something like `if n.offset + i > n.Len() {`?

Reply:
> But also, shouldn't this be something like if n.offset + i > n.Len() {?
🤔 I don't think so, since `n.Len()` factors in the offset.  `n.Len()` returns what the current nibble length of the key it represents.  So you can't advance more than what you have.

Reply:
Ended up writing some tests in dddbecc0a8de8d1d4f91342c296b6b7eb76884c8 to validate.

Reply:
Ah right, thanks!